text/0005-disallow-whitespace-in-entityuid.md

```diff
@@ -0,0 +1,94 @@
+- Date Entered FCP:
+- Date Accepted:
+- Date Landed:
+
+## Summary
+
+Disallow embedded whitespace, comments, and control characters in EntityUID
+
+## Basic example
+
+Currently, the following syntax is legal:
+```
+permit( 
+  principal == ExampleCo :: Photoflash  ::  //This is a comment
+       :: User::"alice",
+  action,
+  resource
+);
+
+permit( 
+  principal == ExampleCo::Photoflash::User:://"alice"
+
+  "bob",
+  action,
+  resource
+);
+```
+
+This change would modify the parser so that the same policies must be expressed as follows:
+```
+permit( 
+  principal == ExampleCo::Photoflash::User::"alice",
+  action,
+  resource
+);
+
+permit( 
+  principal == ExampleCo::Photoflash::User::"bob",
+  action,
+  resource
+);
+```
+## Motivation
+Similar to other programming languages such as Rust, Cedar is currently whitespace agnostic. This leads to the ability to embed whitespace, control characters like newlines, and comments around the `::` delimiter in an Entity UID.
+
+This capability was little known, even amongst Cedar team members, and was only identified as a result of a typo in a Cedar demo. This behavior presents a risk along a few dimensions. First, much of the tooling around Cedar presumes these values are strings.
+
+Examples:
+1. The Cedar Schema format models the schema configuration under a JSON key for the namespace. By all appearances, this is a JSON string rather than a fragment of Cedar syntax. Policy stores which index schemas by namespace are unlikely to recognize the need to normalize the value, leading to the possibility of storing duplicate schema definitions for "ExampleCo::Photoflash" and "ExampleCo  ::  Photoflash" and indeterminate behavior regarding which schema takes effect at runtime.
+2. Policy Stores can implement logic that relies on string comparisons against the EntityType. In a real issue, an application using Cedar sought to preclude callers from passing Actions in the inline slice of entity definitions. It did so by checking if an EntityType matched `.*::Action`. It presumed that `:: Action` was invalid syntax and would be rejected by the Cedar evalator, the same as any other syntatically invalid input. This resulted in a bug, as it allowed callers to avoid the extra validation that the application sought to enforce. While it is technically possible for the application to resolve this by using Cedar tooling to normalize the value, the little-known nature of this Cedar behavior implies that few will know they should normalize the value. Hence, this is likely to result in bugs in other Cedar-based implementation with similar logic.
+3. Customers are anticipated to build meta-permissions layers that restrict callers to manipulating policy store contents for only certain namespaces. This may lead to policies such as `forbid(...) when {context.namespace = "ExampleCo::Photoflash"};`. There is a risk that an unauthorized actor could bypass this restriction by using a namespace with embedded spaces. 
```

Comment:
If we stick to the tenet where we should avoid using direct string representations or use normalized ones, we'd recommend `namespace` attribute to be a record or a string containing normalized namespaces.

Reply:
@shaobo-he-aws to be clear, you are suggesting that Rust codewriters use the [EntityTypeName](https://github.com/cedar-policy/cedar/blob/main/cedar-policy/src/api.rs#L938) or [EntityNameSpace](https://github.com/cedar-policy/cedar/blob/main/cedar-policy/src/api.rs#L959) structs directly? I.e., when parsing a JSON schema, the parser should invoke the `from_str` methods of these structs, and then when using them as keys in a database, they should invoke `fmt` to output to a normalized string?

That seems doable to me, but I can understand the concern raised in the RFC that developers may not think to do this. Indeed, I think the third example is something that application writers could easily miss.

Reply:
Yes, that's my idea.